src/main/kotlin/org/rust/ide/intentions/MoveTypeConstraintToParameterListIntention.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+class MoveTypeConstraintToParameterListIntention : RsElementBaseIntentionAction<RsWhereClause>() {
+
+    override fun getText() = "Move type constraint to parameter list"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): RsWhereClause? {
+        val whereClause = element.ancestorStrict<RsWhereClause>() ?: return null
+        val wherePredList = whereClause.wherePredList
+        if (wherePredList.isEmpty()) return null
+
+        val typeParameterList = whereClause.ancestorStrict<RsGenericDeclaration>()?.typeParameterList ?: return null
+        val lifetimes = typeParameterList.lifetimeParameterList.mapNotNull { it.quoteIdentifier.text }
+        val types = typeParameterList.typeParameterList.mapNotNull { it.name }
+        if (wherePredList.any { it.lifetime?.quoteIdentifier?.text !in lifetimes && it.typeReference?.text !in types })
```

Comment:
I think it's better to check equality using `resolve` approach - get reference from some `RsReferenceElement`, call `resolve` and use result of this call for equality check